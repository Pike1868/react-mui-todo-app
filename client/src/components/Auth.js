import { useState } from "react";
import { Button } from "@mui/material";
import { useCookies } from "react-cookie";

const Auth = () => {
  const [cookies, setCookie, removeCookie] = useCookies("");
  const [isLogIn, setIsLogin] = useState(true);
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [confirmPassword, setConfirmPassword] = useState("");

  const [error, setError] = useState(null);

  const viewLogin = (status) => {
    setError(null);
    setIsLogin(status);
  };

  const handleSubmit = async (e, endpoint) => {
    e.preventDefault();

    if (!email || !password) {
      setError("Email and password are required");
      return;
    }

    if (!isLogIn && password !== confirmPassword) {
      setError("Passwords must match");
      return;
    }
    const response = await fetch(
      `${process.env.REACT_APP_SERVER_URL}/${endpoint}`,
      {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ email, password }),
      }
    );

    const data = await response.json();

    if (data.detail) {
      setError(data.detail);
    } else {
      setCookie("Email", data.email);
      setCookie("AuthToken", data.token);
    }
  };
  return (
    <div className="auth-container">
      <div className="auth-container-box">
        <form onSubmit={(e) => handleSubmit(e, isLogIn ? "login" : "signup")}>
          <h2>{isLogIn ? "Please Log In" : "Please Sign Up"}</h2>
          <input
            type="email"
            placeholder="email"
            onChange={(e) => setEmail(e.target.value)}
          />
          <input
            type="password"
            placeholder="password"
            onChange={(e) => setPassword(e.target.value)}
          />
          {!isLogIn && (
            <input
              type="password"
              placeholder="confirm password"
              onChange={(e) => setConfirmPassword(e.target.value)}
            />
          )}
          <input type="submit" />
        </form>
        {error && (
          <p
            style={{
              color: "#fff",
              backgroundColor: "#fb0e0e8f",
              padding: "0.5rem 0",
              borderRadius: "1rem",
              textAlign: "center",
              margin: "2rem 1rem 0",
            }}
          >
            {error}
          </p>
        )}
        <div className="auth-options">
          <Button variant="outlined" onClick={() => viewLogin(false)}>
            Sign Up
          </Button>
          <Button variant="outlined" onClick={() => viewLogin(true)}>
            Log In
          </Button>
        </div>
      </div>
    </div>
  );
};

export default Auth;
