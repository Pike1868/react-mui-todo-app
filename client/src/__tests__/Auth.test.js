// Auth.test.js
import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Auth from "../components/Auth";
import fetchMock from "jest-fetch-mock";

fetchMock.enableMocks();

beforeEach(() => {
  fetch.resetMocks();
});

// Smoke Test
it("renders without crashing", () => {
  render(<Auth />);
});

// Snapshot Test
it("matches snapshot", () => {
  const { asFragment } = render(<Auth />);
  expect(asFragment()).toMatchSnapshot();
});

// Interaction Test: Input Fields
it("allows entering email and password", () => {
  render(<Auth />);

  fireEvent.change(screen.getByPlaceholderText("email"), {
    target: { value: "test@example.com" },
  });
  fireEvent.change(screen.getByPlaceholderText("password"), {
    target: { value: "password" },
  });

  expect(screen.getByPlaceholderText("email").value).toBe("test@example.com");
  expect(screen.getByPlaceholderText("password").value).toBe("password");
});

// Interaction Test: Form Submission
it("submits the form with email and password", async () => {
  fetch.mockResponseOnce(
    JSON.stringify({ email: "test@example.com", token: "fake_token" })
  );

  render(<Auth />);

  // Simulate user input
  fireEvent.change(screen.getByPlaceholderText("email"), {
    target: { value: "test@example.com" },
  });
  fireEvent.change(screen.getByPlaceholderText("password"), {
    target: { value: "password" },
  });

  // Simulate form submission
  fireEvent.submit(screen.getByRole("button", { name: /submit/i }));

  // Check if fetch was called with correct arguments
  expect(fetch).toHaveBeenCalledWith(
    `${process.env.REACT_APP_SERVER_URL}/login`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ email: "test@example.com", password: "password" }),
    }
  );
});
