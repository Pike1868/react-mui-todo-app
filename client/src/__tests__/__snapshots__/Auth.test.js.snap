// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="auth-container"
  >
    <div
      class="auth-container-box"
    >
      <form>
        <h2>
          Please Log In
        </h2>
        <input
          placeholder="email"
          type="email"
        />
        <input
          placeholder="password"
          type="password"
        />
        <input
          type="submit"
        />
      </form>
      <div
        class="auth-options"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Sign Up
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Log In
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
